File: model/status.go
package model

import (
	"time"
)

type Status struct {
	ServiceName string    `json:"service"`
	Version     string    `json:"version"`
	Runtime     string    `json:"runtime"`
	BuildTime   time.Time `json:"buildTime"`
	StartTime   time.Time `json:"start"`
	SystemTime  time.Time `json:"time"`
	Uptime      string    `json:"uptime"`
}

func (z *Status) PopulateNow() {
	z.Populate(time.Now().Truncate(time.Second).UTC())
}

func (z *Status) Populate(now time.Time) {
	z.SystemTime = now
	z.Uptime = z.SystemTime.Sub(z.StartTime).String()
}

File: ui/src/stores/status.ts
import { action, observable } from "mobx";
import { IServerStatus } from "../model/status";
import { to } from "./util";

const blankServerStatus = {
    service: "",
    version: "",
    runtime: "",
    buildTime: new Date(0),
    start: new Date(0),
    time: new Date(0),
    uptime: "",
};

export class StatusStore {
    @observable public client = "";
    @observable public server = blankServerStatus;

    @action public refresh = () => {
        this.updateClientVersion();
        this.updateServerStatus();
    }

    @action public setServerStatus(serverStatus: IServerStatus) {
        this.server = serverStatus;
    }

    @action public updateClientVersion = () => {

        const versionTag =
            Array.from(document.getElementsByTagName("meta"))
                .find((tag: HTMLMetaElement) => tag.getAttribute("name") === "version");

        this.client = versionTag ? versionTag.getAttribute("content") || "UNKNOWN" : "MISSING";

    }

    @action public updateServerStatus = async () => {

        const [rsp, errRsp] = await to(fetch("/api/status", {
            cache: "no-cache",
            credentials: "same-origin",
            method: "GET",
            mode: "same-origin",
        }));

        if (errRsp) {
            // TODO: log error back to server
            console.warn("get server status error", errRsp); // tslint:disable-line:no-console
            return blankServerStatus;
        }

        const serverStatus = await rsp.json();
        this.setServerStatus(serverStatus);

    }

}
